### PR TITLE
fixing rmsprop bug

### PR DESCRIPTION
**Pull request by karpathy**
The current [RMSProp implementation](https://github.com/torch/optim/blob/master/rmsprop.lua) has a critical bug in the following line:

```lua
x:add(-lr, state.tmp)
```

I think the code was trying to **multiply** the learning rate to the `state.tmp` variable and then add it to `x` to perform the parameter update, but right now it's just adding them and adding the result to `x`. I noticed this because my RNN wasn't training with RMSProp at all, when this worked very nicely in my near equivalent Python codebase. 

This pull request fixes the RMSProp implementation and cleans up the code/hyperparams. RMSProp normally has three parameters:

- The learning rate
- The decay rate (`alpha`)
- The smoothing term (`epsilon`)

I fixed the code to contain exactly these three parameters. I also adjusted the code to look similar to what `adagrad.lua` looks like, except the running sum of squares is leaky in RMSProp. I also set the default hyperparameters to values that normally work for me in practice, at least when training RNNs.

### CODE REVIEW

**Comment by soumith** [comment]
Just want to point out that `x:add(-lr, state.tmp)` does: `x = x + (-lr * state.tmp)` using this overload of the function add:
https://github.com/torch/torch7/blob/master/doc/maths.md#res-torchaddres-tensor1-value-tensor2
So it is perfectly valid and correct syntax.

I guess @koraykv will look at the rest of the PR

**Comment by karpathy** [comment]
Doh, I read the function name `add`, and then when I went to the docs I looked at the wrong overload. Sorry about that.

I also jumped to the conclusion because I am used to seeing RMSProp make very fast progress on RNNs and I wasn't seeing this on my Torch model. And I noticed that the RMSProp implementation is somewhat new (24 days) and has no test case in `test/` folder so putting this all together I thought maybe a bug slipped by.

With the version of RMSProp in this PR my RNN loss converges rapidly similar to what I'm used to. It's not clear why my version works for me right away but the current one left the loss flat. I'll stare at it more.

Given that this isn't a bug it's probably best to leave the code alone, preserve backwards compatibility and close the PR.

File: rmsprop.lua [hunk 1]
```diff
@@ -25,27 +22,24 @@ function optim.rmsprop(opfunc, x, config, state)
...
-    -- (3) calculate new mean squared values
+    -- (3) calculate new (leaky) mean squared values
     state.m:mul(alpha)
-    state.m:addcmul(1.0-alpha,dfdx,dfdx):add(epsilon2)
+    state.m:addcmul(1.0-alpha,dfdx,dfdx)
 
     -- (4) perform update
-    state.tmp:copy(state.m):pow(-0.5):clamp(min_gain, max_gain)
-    x:add(-lr, state.tmp)
+    state.tmp:copy(state.m):sqrt()
```
Comment (koraykv) [inline rmsprop.lua]: you could just do state.tmp:sqrt(state.m) to avoid extra copy

**Comment by koraykv** [comment]
I am inclined to pull this one since it gets rid of 3 extra parameters. I am not sure if they are important, it is my bad that I did not check this code when first committed. Since rmsprop is fairly new, I think it is ok to break compatibility. If there are no objections, we should go ahead and merge this.

File: rmsprop.lua (hunk 1 above)
Comment (karpathy) [inline rmsprop.lua]: Thanks, I saw the `:copy()` pattern in all the other scripts (e.g. Adagrad, Adam, etc.) so I assumed it was an intentional. Incorporated.
